### DISCUSSION TITLE
Missing TypeCode for Half, Int128, UInt128?

### DISCUSSION THREAD

**Discussion by msedi** [dotnet/runtime#81150] [Q&A]
I know that the `Half, Int128`, etc. have not been part of the original type system, but are there any plans to add them to the `TypeCode` enum?

Its just a question, originally I was needing them to create generic math. I have to dig a bit if I can resolve it now since the `static interface`  and the "INumeric" methods exist now.

**Comment by Joe4evr** [comment]
> are there any plans to add them to the `TypeCode` enum?

Likely never. Firstly, there may still be many consumers (even indirectly) relying on the fact that `TypeCode` hasn't expanded in two decades (even though I don't think it was ever guaranteed to never expand), so adding new values *now* could have a much larger impact than you think.

Secondly, what `TypeCode` was originally designed for seems like a mistake in modern .NET development. It'd be better if new code doesn't use it at all.

**Reply by MarkPflug** [reply]
> It'd be better if new code doesn't use it at all.

Oh no. I've used this so many times in my code: `switch(typeCode)`.

**Reply by msedi** [reply]
@MarkPflug: Yes, that's what I did too (I thought for performance reasons), but honestly I haven't done recent measurements if this

```
public void SomeMethod(object o)
{
  if (o.GetType() == typeof(int))
  {
  }
}
```

performs the same way. Switiching on `Type`s would be easier, though.

**Reply by msedi** [reply]
@Joe4evr. Thanks for pointing this out. I answered also to @MarkPflug. Do you have experience whether switching on TypeCode or doing it via GetType() (see above) has a performance difference?

**Reply by MarkPflug** [reply]
Indeed, I always did it for what I assumed was the perf-gain of a switch statement for the common type cases over an if/then cascade that would be required with type comparisons. Only in the `Object/default:` case would I `if` for handling things like Guid or other common data types.

If there is a better way to handle this in "modern" code, I'd love to hear it.

**Reply by tannergooding** [reply]
> performs the same way. Switiching on Types would be easier, though.

You're better off doing:
```csharp
public T Add<T>(T x, T y)
{
    if (typeof(T) == typeof(int))
    {
        return (T)(object)(((int)(object)(x)) + ((int)(object)(y)));
    }
    else 
    {
        // ...
    }
}
```

This is fully optimized by RyuJIT for value types and leads to correct/ideal codegen:
* `typeof(T) == typeof(...)` where `...` is some value type is treated as a constant true/false when `T` is a value type
* `(T)(object)(x)` and `(...)(object)(y)` for `T y` and `... x` (again for value types) is optimized to be a direct access of the value, no boxing involved (for reference types, casting to/from object is already "free")

If you can utilize the new generic math features, that's even better as you can write something like:
```csharp

public T Add<T>(T x, T y)
    where T : INumber<T>
{
    return x + y;
}
```

**Reply by tannergooding** [reply]
The latter "generic math features" extends to "static abstracts in interfaces" in general and allows you to avoid all the complexity even when static methods are involved.

**Reply by MarkPflug** [reply]
That makes sense when there is a `T`. I find that I need to use this pattern occasionally with "untyped" (object) data. I'd assume using TypeCode is still a minor win in that case, since there isn't anything for the JIT to specialize on?

**Reply by tannergooding** [reply]
Someone would need to measure. Getting the `TypeCode` is not "cheap" and so may be more expensive than simply comparing the type handle.

Particularly given that getting the `TypeCode` itself involves a type check in the form of `if (value is null) { } else if (value is IConvertible tmp) { tmp.GetTypeCode(); } else { }`

and so just doing `if (obj is int i32) { } else if (obj is long i64) { } else { }` may be better (and also has some C# language/pattern support for things like `switch expressions` iirc).

**Reply by MarkPflug** [reply]
> Someone would need to measure.

Okay: https://gist.github.com/MarkPflug/86d5733e6bee9411761fb02e4c88d874

|          Method |      Mean |    Error |   StdDev |
|---------------- |----------:|---------:|---------:|
|   TypeCodeBench |  46.54 ns | 0.957 ns | 1.245 ns |
| TypeEqualsBench | 132.40 ns | 1.305 ns | 1.090 ns |
|     IsTypeBench |  42.31 ns | 0.459 ns | 0.429 ns |

I'm sticking with the TypeCode switch pattern. Even though it is slightly slower in this benchmark, the `if(obj is Type)` pattern is sensitive to the ordering of the conditions; meaning you should put the common types first. I was surprised at how much slower the `obj.GetType() == typeof(string)` (TypeEqualsBench) was. I assumed it would be roughly the same as the `is Type` approach.

**Reply by msedi** [reply]
@MarkPflug. Thanks for creating the tests, I ran your tests on my machine and interestingly got results
 far more different than yours:

|          Method |      Mean |    Error |   StdDev | Allocated |
|---------------- |----------:|---------:|---------:|----------:|
|   TypeCodeBench | 175.64 ns | 3.263 ns | 3.758 ns |         - |
| TypeEqualsBench | 207.96 ns | 1.944 ns | 1.723 ns |         - |
|     IsTypeBench |  90.73 ns | 1.819 ns | 1.946 ns |         - |

Now the question comes up why this is the case. Here's a bit more runtime info:

```
// Benchmark Process Environment Information:
// Runtime=.NET 7.0.2 (7.0.222.60605), X64 RyuJIT AVX
// GC=Concurrent Workstation
// HardwareIntrinsics=AVX,AES,PCLMUL,POPCNT VectorSize=128
// Job: DefaultJob
```

**Reply by tannergooding** [reply]
> I was surprised at how much slower the obj.GetType() == typeof(string) (TypeEqualsBench) was

This is probably just a case we're missing coverage around type checks. `GetType()` involves an additional method call that `is` does not and so the JIT would need extra work to "see through" this call.

If you want to open an issue we can see about maybe improving it.

**Reply by MarkPflug** [reply]
Must be a difference of the CPU that we're running? Not sure.
```
BenchmarkDotNet=v0.13.4, OS=Windows 10 (10.0.19044.2486/21H2/November2021Update)
Intel Core i7-7700K CPU 4.20GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET SDK=7.0.102
  [Host]     : .NET 7.0.2 (7.0.222.60605), X64 RyuJIT AVX2
  DefaultJob : .NET 7.0.2 (7.0.222.60605), X64 RyuJIT AVX2
```

**Reply by tannergooding** [reply]
> Now the question comes up why this is the case

It's potentially due to the scenario being tested/considered.

Proper `switches` (that is those that compile down to `switch` in IL, and not those that compile down to `if/else` chains) have historically limited inlining and a few other optimizations.

We've gotten significantly better about that in the latest releases, but its possible some edge cases still remain.

**Reply by MarkPflug** [reply]
TypeCode should reduce to a jump table switch right, because it is "dense"?

**Reply by msedi** [reply]
@MarkPflug. For sure, this is mine

```
BenchmarkDotNet=v0.13.4, OS=Windows 10 (10.0.19045.2486)
Intel Xeon CPU E5-2670 0 2.60GHz, 2 CPU, 32 logical and 16 physical cores
.NET SDK=7.0.200-preview.22628.1
  [Host]     : .NET 7.0.2 (7.0.222.60605), X64 RyuJIT AVX
  DefaultJob : .NET 7.0.2 (7.0.222.60605), X64 RyuJIT AVX
```

I would agree, that there is a difference in execution time as an offset, but I cannot explain the factor of ~2 on my side between `TypeCodeBench` and `IsTypeBench` (@tannergooding: Do you have an idea what is the cause of this?)

**Reply by msedi** [reply]
Sorry, I don't want to flood you, but I did one more thing on Marks' benchmark. I have added

`[SkipLocalsInit, MethodImpl(MethodImplOptions.AggressiveOptimization | MethodImplOptions.AggressiveInlining)]
`

`AggressiveOptimization` doesn't do much on my machine, `SkipLocalsInit` already has "better" results. But the major part comes from `AggressiveInlining`.

|          Method |     Mean |    Error |   StdDev | Allocated |
|---------------- |---------:|---------:|---------:|----------:|
|   TypeCodeBench | 45.53 ns | 0.896 ns | 0.920 ns |         - |
| TypeEqualsBench | 23.59 ns | 0.503 ns | 0.471 ns |         - |
|     IsTypeBench | 57.27 ns | 1.164 ns | 1.245 ns |         - |

**Reply by tannergooding** [reply]
Could be a few different factors, its hard to tell without looking at the disassembly and/or JIT dump

It could also be a hardware specific difference (such as in supported instruction sets which impacts downstream codegen).

For BDN the easiest thing is https://benchmarkdotnet.org/articles/features/disassembler.html. But you can also try out https://github.com/EgorBo/Disasmo or a few other mechanisms

**Reply by SetTrend** [reply]
Actually I was just about asking a similar question here … I'm often using `TypeCode` to save dynamic data in databases as strings.

```sql
CREATE TABLE Values
( Id INT PRIMARY KEY
, TypeCode INT NOT NULL
, ValueText NVARCHAR(1000) NOT NULL
)
```

This way it's very easy to save data to or read it from the database:

```c#
using System;

internal record Value<T>
{
  internal int Id { get; set; }
  internal TypeCode TypeCode { get; set; }
  internal string ValueText { get; set; }

  internal T Val => (T)Convert.ChangeType(this.ValueText, this.TypeCode);

  internal Value(T val)
  {
    this.TypeCode = Convert.GetTypeCode(val);
    this.ValueText = Convert.ToString(val);
  }
}
```

As such, I'd very much endorse getting [`System.TypeCode`](https://learn.microsoft.com/en-us/dotnet/api/system.typecode) (and, thus, [`System.Convert`](https://learn.microsoft.com/en-us/dotnet/api/system.convert)) extended with the new intrinsic base data types, like, e.g., [`System.DateOnly`](https://learn.microsoft.com/en-us/dotnet/api/system.dateonly).

**Reply by msedi** [reply]
@SetTrend. For my purpose things have a little cleaned up since there is the `IParseable` static interface. Nevertheless there is currently no good way of making things type "independent", meaning that you don't have to treat each type separately. As far as I know this can only be done via reflection `GetType() -> Invoke` and so on (which creates its own problems, AOT, performance, etc).

I think this is why there is so much focus on source generators since reflection would prevent so many new things. Nevertheless how it currently works with generics is a bit of a plague. I have so much code doing something stupid repetitive code which is error prone and too much work for such little things like:

  ```cs
 public static T ReadLittleEndian<T>(ReadOnlySpan<byte> source)
   {
       if (typeof(T) == typeof(ushort))
           return (T)(object)BinaryPrimitives.ReadUInt16LittleEndian(source);
       else if (typeof(T) == typeof(short))
           return (T)(object)BinaryPrimitives.ReadInt16LittleEndian(source);
       else if (typeof(T) == typeof(uint))
           return (T)(object)BinaryPrimitives.ReadUInt32LittleEndian(source);
       else if (typeof(T) == typeof(int))
           return (T)(object)BinaryPrimitives.ReadInt32LittleEndian(source);
       else if (typeof(T) == typeof(ulong))
           return (T)(object)BinaryPrimitives.ReadUInt64LittleEndian(source);
       else if (typeof(T) == typeof(long))
           return (T)(object)BinaryPrimitives.ReadInt64LittleEndian(source);
       else if (typeof(T) == typeof(Half))
           return (T)(object)BinaryPrimitives.ReadHalfLittleEndian(source);
       else if (typeof(T) == typeof(float))
           return (T)(object)BinaryPrimitives.ReadSingleLittleEndian(source);
       else if (typeof(T) == typeof(double))
           return (T)(object)BinaryPrimitives.ReadDoubleLittleEndian(source);

       throw new NotSupportedException($"{typeof(T)} is not supported.");
   }
```

If you look in [source.dot.net ](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Buffers/Binary/BinaryPrimitives.ReadLittleEndian.cs,84880e9510ac88ac), I'm wondering why MS hasn't come up with as better solution. Obviously they like to write a lot of code ;-)

In my opinion there are three problems:

1. lack for constraint support (which can create runtime exceptions only, I cannot check it beforehand) and it seems that I cannot limit the method doing this:

```cs
      public static T ReadLittleEndian<T>(ReadOnlySpan<byte> source) 
          where T : IBinaryFloatingPointIeee754<float>, IBinaryFloatingPointIeee754<Half>...
```

the next would be much better, although it does not grow when adding new type (as has happend with Half or Float64, Long 128), but you would recognize immediately during compile time and not when it happens during runtime.

```cs
      public static T ReadLittleEndian<T>(ReadOnlySpan<byte> source) 
          where T : float or Half
```

2. Even if there would by a support I still need to distingush the individual types
3. The lack of creating a new type, the `new constraint` is too limited, and if it doesn't fit you always have to go with the `Activator` and `MakeGenericType` which slows down things tremendously.

So it seems we have to write out code until there is some better support for this.

**Reply by Joe4evr** [reply]
@SetTrend [This is the energy you're giving off.](https://i.imgur.com/9ZJkPvV.mp4)

**Reply by SetTrend** [reply]
@Joe4evr: &nbsp; Let's not get off-topic.

Sure, there are people creating `Settings` database tables with a single row and 300<sup>++</sup> nullable columns, mainly `null`. And whenever a new application setting is implemented, they are migrating their database. They refrain from normalizing databases.

I prefer the more versatile, dynamic approach.

**Reply by tannergooding** [reply]
There are no plans to version `TypeCode` or `IConvertible` in the future.

Developers should use features like `where T : INumber<T>` and then `T.CreateChecked`, `T.CreateSaturating`, and `T.CreateTruncating` where possible instead.

Similarly, if they need to do a type specific behavior and specialize, the recommendation is to use `if (typeof(T) == typeof(...))` or `if (obj is ... instance)` (replacing `...` with the concrete type they're specializing for).

----------------------------

> If you look in [source.dot.net ](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Buffers/Binary/BinaryPrimitives.ReadLittleEndian.cs,84880e9510ac88ac), I'm wondering why MS hasn't come up with as better solution. Obviously they like to write a lot of code ;-)

First, a lot of this code existed prior to features like generic math. But more importantly, this isn't the kind of code that can or should be shared generically because what you need to do (especially to keep it performant) is unique per type. We do simplify internal implementations where it is possible/relevant.

In the case of your code, you're reading a different number of bytes per `T`, so the shape itself is one that the BCL would likely not expose; instead we would require it return the number of bytes read as an additional output.

This is notably also why `TBinaryInteger.ReadLittleEndian` reads the entire span and treats the span as a single two's complement number, because when doing such operations generically, you need to support say `int` which might always write 4-bytes being able to read something like `long` which always writes 8-bytes even if the value is something like `0`.

> lack for constraint support (which can create runtime exceptions only, I cannot check it beforehand) and it seems that I cannot limit the method doing this:

.NET generics are not C++ templates, they need to persist in metadata and follow some ABI contract that can be understood at JIT compile time.

This means you can't just write arbitrary code and expect things to work and only fail if something mismatches. You need to have some kind of common "contract" (i.e. the constraint) against which calls can be emitted so it knows how to resolve say `ReadLittleEndian`.

> The lack of creating a new type, the new constraint is too limited, and if it doesn't fit you always have to go with the Activator and MakeGenericType which slows down things tremendously.

Activator.CreateInstance is relatively fast, especially in recent releases.

But, there's also a simple solution to this problem as you can define some `interface ICreatable<T> { static abstract T Create(); }` and have your types implement it. If you need it to work more like a factory pattern, you have many other choices as well which allows you to then provide your own factory that creates instances of type `T` (just take in `TFactor, T` as the type parameters).

**Reply by msedi** [reply]
> Developers should use features like where T : INumber<T> and then T.CreateChecked, T.CreateSaturating, and T.CreateTruncating where possible instead.

I agree that things have changed recently and a lot of things are getting better with static abstract interfaces.

> First, a lot of this code existed prior to features like generic math. But more importantly, this isn't the kind of code that can or should be shared generically because what you need to do (especially to keep it performant) is unique per type. We do simplify internal implementations where it is possible/relevant.

Also here I agree, I have the same implementations as you do for different data types, but the lack of constraining T to some very specific type is a big problem. For example the MKL can only do `float` and `double`, but there is not chance to limit it to these two, right? I could constrain it maybe to `INumber<float..double>` but this means I can use everything that is an `INumber<float>`, but I only can only deal with float/double array, so this opens the door for other problems. Am I right?

> In the case of your code, you're reading a different number of bytes per T, so the shape itself is one that the BCL would likely not expose; instead we would require it return the number of bytes read as an additional output.
> 
> This is notably also why TBinaryInteger.ReadLittleEndian reads the entire span and treats the span as a single two's complement number, because when doing such operations generically, you need to support say int which might always write 4-bytes being able to read something like long which always writes 8-bytes even if the value is something like 0.
I'm not sure if I got this. What is wrong with `public static T ReadLittleEndian<T>(ReadOnlySpan<byte> source)`. Sure, I could also live with `public static bool TryReadLittleEndian<T>(ReadOnlySpan<byte> source, out T result)` but ultimately you can read

`ReadLittleEndian<int> as ReadInt32LittleEndian and ReadLittleEndian<float> as ReadSingleLittleEndian`

I can see no difference in this. Additionally when it comes to the length 

`public static uint ToUInt32(ReadOnlySpan<byte> value)`

and it siblings, they already check the length. So I agree the BitConverter exists since .NET Framework 1.1 (I think), but there really is a whole lot of code that is truly repetitive. There are of course things that are different, but for the end user the problem is when you need to scale between different datatypes in your application, lets say for example there are conditions (data size of the input data) where you need to switch from double to float to half a better generic support without having to write each code path. I understand the situation in the BCL and in our product code we have the same troubles an implement the optimized code path invidually sometimes really going down to ASM. Also when using CUDA you need to adjust your code pathes also for every hardware, so I'm really aware of the problems. But when it comes to higher level programming things should resolve to easier code. I'm not sure if you DICOM. This standard has nowadays 30 different datatypes which translate to DateOnly, TimeOnly, DateTime, TimeTimeOffset, string in 6-7 different flavors, float, double float-string, double-string and so on, additionally with different endianness, dependent on their encoding, sometimes switching the encoding and/or endianness in the same stream from one item to the other. With a few tweaks of the BCL things are perfectly possible, but then a Half type is introduced by the BCL, the new type was not in a unit test before and suddenly everything breaks because there was no contraint that it has not been implemented before, sure there is an exception that the type is not supported, but I 1. wasn't tested since the test didn't know about it and 2. it is too dynamic to cover all possibilities. All this is causing a lot of pain and work on our side.

> .NET generics are not C++ templates, they need to persist in metadata and follow some ABI contract that can be understood at JIT compile time.
Sure, which I like, because C++ templates have their own problems ;-) Nevertheless is the lack of constraining generics to some special type a problem of the metadata then?

> Activator.CreateInstance is relatively fast, especially in recent releases.
I tried, but still our implementation seems to be better. I can try again, but the problem is that it is hard to tell when things have improved and when they have improved. It would be good if there would be some table that shows some performance trend for selected implementations/features/libraries/methods. I think the BCL already has benchmarks. Can they be tabulated and put somewhere so that the community can follow the trends and do some lookups there?

> But, there's also a simple solution to this problem as you can define some interface ICreatable<T> { static abstract T Create(); }
This does not always work since we don't always own the used libraries. Additionally, the problem with T is that you do not always have a T but a Type, then ou need to go with MakeGenericType, etc and then the performance gets down.  I try to get around this as much as possible, but sometimes you have to. The last time I had to do it I want to expression trees and dynamic runtime comilation to improve performance.

Sorry for too much text ;-) I would sometimes prefer a discussion in personal because the topic is too complex to describe this in this limited environment.

**Reply by tannergooding** [reply]
The typical way to handle supporting just two types like `float` and `double` is to have just those two for a public signature and have them defer to an internal shared implementation that is `where T : IBinaryFloatingPointIeee754<T>`

This allows you to correctly express the supported set to the consumer without forcing you to duplicate code across all supported types.

You can optionally expose a generic overload as well, but it then needs to check that T is only one of the supported types (like `Vector128<T>` does)
